product_barcode_required/models/product_barcode_mixin.py

```diff
@@ -0,0 +1,40 @@
+# Copyright 2020 Camptocamp SA
+# License LGPL-3.0 or later (https://www.gnu.org/licenses/lgpl)
+
+from odoo import _, api, exceptions, fields, models
+
+
+class BarcodeRequiredMixin(models.AbstractModel):
+    _name = "product.barcode.required.mixin"
+    _description = "Barcode required mixin"
+
+    is_barcode_required = fields.Boolean(compute="_compute_is_barcode_required")
+
+    @api.onchange("default_code")
+    def _onchange_code(self):
+        for rec in self:
+            if rec.is_barcode_required and rec.default_code:
+                rec.barcode = rec.default_code
+
+    @api.depends("type", "barcode")
+    def _compute_is_barcode_required(self):
+        for rec in self:
+            rec.is_barcode_required = (
+                rec._is_barcode_required_enabled() and rec._is_barcode_required()
+            )
+
+    def _is_barcode_required(self):
+        self.ensure_one()
+        return self.type != "service" and not self.barcode
+
+    def _is_barcode_required_enabled(self):
+        return self.env.company.product_variant_barcode_required
+
+    def _check_barcode_required(self):
```

Comment:
@simahawk Why not `api.constraints("barcode")`

Reply:
because I don't want to enable it by default. Also, on template is useless since it comes from the 1st variant.

Reply:
In such a case you could add the constrains only into 'product.product'. That will avoid to override `write`and `create` in `product.template` and `product.product`, will remove a lot of code and avoid using the context to pass arguments (IMO it's always a bad practice to use the context in this way even if in > 13, there is no more performance impact when using this kind of trick) 

```python
class ProductProduct(models.Model):
    _name = "product.product"
    _inherit = ["product.product", "product.barcode.required.mixin"]

    @api.constrains("is_barcode_required")
    def _chek_is_barcode_required(self):
        return super()._chek_is_barcode_required(self):

class ProductTemplate(models.Model):
    _name = "product.template"
    _inherit = ["product.product", "product.barcode.required.mixin"]
```

Reply:
you cannot avoid the ctx key in the template because the template triggers the create/write of product.product when you update variant attributes. BUT at that stage you might have no barcode because the one that you set on the template is applied only to the 1st variant.
Also, to limit the possibility to have errors, I added a call to the onchange in the create/write to make sure we can have a default barcode when possible.

Reply:
ah, what I could change is to set the key only in `_create_variant_ids`.

Reply:
In the same time `barcode` could become a computed field `readonly=False`. That would avoid to have to call to `_onchange_code` into the `create` and `write`method 

```python
class ProductProduct(models.Model):
    _name = "product.product"
    _inherit = ["product.product", "product.barcode.required.mixin"]

    barcode = fields.Char(compute="_compute_barcode", readonly=False)
 
    @api.depends("default_code", "is_barcode_required")
    def _compute_barcode(self):
        for rec in self:
            if rec.is_barcode_required and rec.default_code:
                rec.barcode = rec.default_code
```
Do we even still need the mixin? `

The code for `is_barcode_required` field could be put into `product.product` and the field makes available into `product.tempate` as related

```python
class ProductProduct(models.Model):
    _inherit = "product.product"
    
    barcode = fields.Char(compute="_compute_barcode", readonly=False)
    is_barcode_required = fields.Boolean(compute="_compute_is_barcode_required")
 
    @api.depends("default_code", "is_barcode_required")
    def _compute_barcode(self):
        for rec in self:
            if rec.is_barcode_required and rec.default_code:
                rec.barcode = rec.default_code

    @api.depends("type", "barcode")
    def _compute_is_barcode_required(self):
        for rec in self:
            rec.is_barcode_required = (
                rec._is_barcode_required_enabled() and rec._is_barcode_missing()
            )

    def _is_barcode_missing(self):
        self.ensure_one()
        return self.type != "service" and not self.barcode

    def _is_barcode_required_enabled(self):
        return self.env.company.product_variant_barcode_required
    
    @api.constrains("is_barcode_required")
    def _chek_is_barcode_required(self):
        """Check if barcode required, to be used in your own contraint."""
        # Make error nicer up to 30 records.
        failing = [x.display_name for x in self[:30] if x.is_barcode_required]
        if failing:
            failed_list = "\n  * " + "\n  * ".join(failing)
            raise exceptions.ValidationError(
                _("These products have no barcode:\n{}").format(failed_list)
            )

class ProductTemplate(models.Model):
    _inherit = "product.template"

    is_barcode_required = fields.Boolean(related="product_variant_id.is_barcode_required")
    
```

Reply:
barcode computed: read my https://github.com/OCA/product-attribute/pull/677#discussion_r510658087 ;)
mixin: yes, because you want to check on the template as well. Also, you could reuse it for other models having barcodes to make required.
constrains: it does not work as long as is_barcode_required does not change.